ykcompile/src/lib.rs

```diff
@@ -43,6 +43,130 @@ lazy_static! {
     static ref PTR_SIZE: u64 = u64::try_from(mem::size_of::<usize>()).unwrap();
 }
 
+macro_rules! binop_as {
+    ($name: ident, $op:expr) => {
+        fn $name(&mut self, dest: &IPlace, size: u64, opnd2: &IPlace) {
+            let src_loc = self.iplace_to_location(opnd2);
+            match src_loc {
+                Location::Register(r) => match size {
+                    1 | 2 | 4 => todo!(),
+                    8 => {
+                        dynasm!(self.asm
+                            ; $op Rq(*TEMP_REG), Rq(r)
+                        );
+                    }
+                    _ => unreachable!(format!("{}", SIR.ty(&dest.ty()))),
+                },
+                Location::Mem(..) => todo!(),
+                Location::Const { val, .. } => {
+                    let val = val.i64_cast();
+                    if val > i64::from(u32::MAX) {
```

Comment:
```suggestion
                    if val > i64::from(u32::MAX) || val < 0 {
```

Reply:
I think this would lead to `-3u8` also ending up in this branch. Did you maybe mean:
```rust
if val > i64::from(u32::MAX) || val < -i64::from(u32::MAX) {
```

Reply:
or just
```rust
if abs(val) > i64::from(u32::MAX) {
```

Reply:
When it comes to things like `-i64::from(u32::MAX)`, it's worth knowing that `(uX:MAX)` and `uX:MIN` are not symmetric in twos complement.

Reply:
I think what @bjorn3 is saying is that if the number is negative then by definition its 64-bit representation won't fit in a 32-bit register. In signed integers the top bit has to be `1` for negative numbers.

I'm starting to dislike this signed/unsigned malarky here. We don't actually care about the signedness of the data in `val` here, only the bit pattern.

For this check, we just want to know if any higher-order bits are set. We can take signedness out of the picture with:
```
if val >> 32 != 0 { ... }
```

Reply:
You could also do `u32::try_from(val).is_err()` I think.

Reply:
@vext01 I understood what @bjorn3 was saying, but in this case `val` can be a negative `i8`, e.g. `-3u8`, which we then cast to a `i64` which is `< 0`. We then end up in this `todo()` branch whereas we can just use the normal instructions below.

@bjorn3 `u32::try_from(val).is_err()` sounds like it would work better.

Reply:
Can I ask a dumb question: what are we trying to achieve in this part of the code?

Reply:
We are trying to figure out if the 64-bit constant can fit in a 32-bit register, as in x86_64 you can't move a whole 64-bit constant into a register in one go.

@ptersilie It's right that we hit the todo for -3 as its two's compliment encoding can't fit in 32-bits.

-3i64 is `11111111 11111111 11111111 11111111 11111111 11111111 11111111 11111101`.

Notice that the case for 8-bytes below only actually loads an i32 DWORD. That's because we are sure that the top 32-bits are all zero, so we can load it all in one go.

Reply:
Isn't this something that dynasm does for us? [Obviously I'm a bit clueless!] Otherwise we have to do a lot of bit-checking stuff and it's quite fiddly.

Reply:
Afraid not. It doesn't do any instruction selection for you. I don't know of an assembler that does.

Reply:
Personally I'd be totally fine if we currently `todo` for everything that's not an easy case. I would define easy cases as including, for example, `-1` so it's not *totally* easy, but I'm not sure I care about encoding `u64::MIN` as a constant at the moment.

Reply:
Yep. That's what the todo is doing. We can implement those later, and it's better to implement them when some test actually touches that case anyway.

Reply:
Makes sense!

Reply:
OK, let's see if Lukas agrees.

Reply:
> @ptersilie It's right that we hit the todo for -3 as its two's compliment encoding can't fit in 32-bits.

But we shouldn't hit it for -3i8, which we would do `if val >> 32 != 0` or `val < 0`, since we cast to i64 first, so `-3i8 = 0xfd` becomes `0xfffffffffffffffd`. But I think @bjorn3 suggestion to use `u32::try_from(val).is_err()` should work.